### ISSUE TITLE
Add checksums for yadm encrypt commands

### ISSUE THREAD

**Issue by alexbarcelo** [TheLocehiliosan/yadm#126]
I have a bunch of files with sensible information in them and use the encrypt feature of yadm. It works great so far, but because the files are outside the git, `yadm status` doesn't tell me if I changed them or I didn't.

I was thinking, as a new feature of encrypt, to add checksums. Something along having a file with a list of all the files that are inside the files.gpg, and they md5sum. This would be auto-generated by `yadm encrypt`, and a `yadm encrypt status` (as an example) would say which files have been modified.

Does it make sense? Before starting to work on this (as a PR) I wanted to be sure that this makes sense and that it is not an already-existing feature I missed.

**Comment by TheLocehiliosan** [comment]
@alexbarcelo - This is not an existing feature. Today I would suggest that you consider creating a hook called `.yadm/hooks/post_status`. Creating a hook is an easy way to begin adding features to an already existing **yadm** operation. If the hook works well, it could be the basis of a **yadm** feature addition.

I do have tentative plans to add support for git-crypt which would allow inline encryption of files. Git-crypt implements transparent encryption of data using Git "filters". This results in the encrypted data working like everything else for operations like `git status` or `git diff`. However, this git-crypt support is probably not in the immediate future.

Here are my thoughts on the ideas proposed:

1. Consider creating a hook to test out your ideas. The hooks system allows injecting your own code in a seamless way, and can still be shared with others. I do plan on establishing a list of user contributed hooks once there are a number of them with universal appeal.
2. If you plan to extend **yadm** with this type of feature, try to use test-driven development. I do not plan on merging in any code that is not covered by automated tests. Tests are far easier to write up-front, rather than creating them after development is complete. I'm more inclined to include code that already has robust tests, as adding the tests myself will slow down the inclusion. Indeed, tests are often more effort to write well than the features being tested.
3. Consider holding off until the new testing system is complete. The existing tests are all written with Bats, and over time I've found that testing system to be very difficult to work with. I'm currently in the process of rewriting all of the tests using pytest, which should allow for more sophisticated tests written much quicker. I've made a lot of progress towards migrating all the tests, but that is not complete just yet. This test rewrite is currently a blocker for all my future development with **yadm**, so all of my current development time goes toward that effort. I think I will be done pretty soon.

**Comment by alexbarcelo** [comment]
Thanks for all the feedback!

Right now I have my hands full, but it's great to have those suggestions and general overview handy :) But I can start playing aroung with that hook soon, and then patiently write tests and wait for the new testing system.

**Comment by con-f-use** [comment]
Somewhat related: Currently all the encrypted files are stored in one large encrypted binary blob. If you change one file inside the binary blobb, the whole thing changes and needs to be versioned. Leading to a large repository over time, as these changes accumulate.

To alleviate the situation, each file should be encrypted into a separate file. The encryption should take place if and only if the specific one changed.

Otherwise you have a situation like this:

Encrypted files and  their encrypted sizes in commit one: FileA.txt (1kB), FileB.bin (10MB) --> repo-size: 20.001 MB

Small addition is made to FileA.txt

Encrypted files their encrypted sizes in commit two: FileA.txt (2kB), FileB.bin (10MB) --> repo-size: 40.003 MB (because commit one is stored as well and the whole encrypted blobb changed)

If files were checksummed, tracked and encrypted individually on a need-to basis, the repo-size after commit two, would be 20.003 MB. Software likes SpiderOak even goes one step further and would break the 20MB file up into smaller individually tracked files, so that if only a small part of the big files changes, only that small part would be stored extra.

**Comment by mzuther** [comment]
*(Sorry, posted this in the wrong issue)*

I have just added pull request #188 with two contributed hooks that implement checking of encrypted files with the SHA-512/256 algorithm.  Note: MD5 is **not secure** and should not be used for cryptographic hashes any more.  Feel free to test my PR and report any findings!

**There is a small change in YASM itself, so you'll have to clone https://github.com/mzuther/yadm/tree/encrypt_with_checksums for testing.**

Merry Christmas! :)

**Comment by con-f-use** [comment]
Just as a side-node: While MD5 is (theoretically) not secure in cryptographic applications, it is okay to use when checking fore file integrity and changes. Especially in our setting, where security comes from gpg.

I would also like to see git's LFS feature for actually storing the encrypted files. That would also circumvent the problems outlined in my post above.

**Comment by mzuther** [comment]
> While MD5 is (theoretically) not secure in cryptographic applications, it is okay to use when checking for file integrity and changes. Especially in our setting, where security comes from gpg.

Thanks for clarifying. However, I'm not so sure about this (but I am not a security professional). Even if security comes from GPG, if you can restore the original file from the checksum (preimage attack), GPG's security won't help you.

Over time, MD5 has shown several serious flaws and there is a [preimage attack](https://en.wikipedia.org/wiki/Hash_function_security_summary#Preimage_resistance) (although more of a theoretical nature). And those are only the published attacks. Technology should move away from MD5 as soon as possible, which is why I chose a version of SHA assumed to be secure.

But this is neither here, nor there, and I digress... :)

> I would also like to see [git's LFS feature](https://www.atlassian.com/git/tutorials/git-lfs) for actually storing the encrypted files.

I think you should open a new issue for this. Otherwise your idea/wish might get lost.

**Comment by con-f-use** [comment]
"Preimage attack" does not mean, you can compute the original file just from the hash. That is impossible for files larger than the hash (128 bit for md5), because you lose information by hashing, i.e. for any given hash there are infinite files that have that hash. Rather a successful preimage attack means for a given original file, you have the capability to produce another file that has the same hash as your original faster than brute-forcing it. That is bad because if you want to use the hash to verify that no man in the middle swapped it out with something malicious, now you can't be sure. But since we have encrypted with our private key using gpg, we know an attacker couldn't have swapped it, because they would have to have known our gpg private key, too.

Again, it doesn't hurt to use SHA and I wouldn't change it. I just wanted to clarify a common misconception. It's a bad habit from when I was an educator. Trust me, I'm a security professional now :smile:

**Comment by mzuther** [comment]
> That is impossible for files larger than the hash (128 bit for md5), you lose information by hashing, i.e. for any given hash there are infinite files that have that hash, though almost all of them are quite large.

Thanks! This makes so much sense that I'm now slapping my head because I didn't notice it before.  :smile:

**Comment by TheLocehiliosan** [comment]
Since 2.4.0 yadm has had @mzuther's hook for checksums in the contrib/hooks directory. I'm going to close this issue now.
